# Chapitre 2 : Data Science : Toolbox
Email : <a href='mailto:madani.a@ucd.ac.ma'>madani.a@ucd.ac.ma</a>
<img src='images/data_science.png' width='50%'>
<p>
Dans ce chapitre, nous allons apprendre les bases de la bibliothèque <strong>Pandas</strong> pour les données scientifiques.
</p>
<p>
Commençons par importer les paquets dont nous aurons besoin pour notre programme. Dans la première cellule, nous avons mis le code pour importer la bibliothèque pandas en tant que pd. C'est pour la commodité; chaque fois que nous devons utiliser certaines fonctionnalités de la bibliothèque pandas, nous écrirons pd au lieu de pandas. Nous allons également importer les deux bibliothèques de base mentionnées dans la bibliothèque numpy du chapitre book comme np et la bibliothèque matplotlib.pyplot comme plt.
</p>

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## DataFrames

<p>
La structure de données clé dans Pandas est l'objet <strong>DataFrame</strong>. Un DataFrame est essentiellement une structure de données tabulaire, avec des lignes et des colonnes. Les lignes ont un index spécifique pour y accéder, qui peut avoir n'importe quel nom ou n'importe quelle valeur. Dans Pandas, les colonnes sont appelées <strong>Series</strong>, un type spécial de données, qui consiste essentiellement en une liste de plusieurs valeurs, où chaque valeur a un index. Par conséquent, la structure de données DataFrame peut être considérée comme une feuille de calcul (qui ressemble à une feuille Excel), mais elle est beaucoup plus flexible.
</p>

## Créer un nouveau DataFrame

<p>
Pour comprendre comment cela fonctionne, voyons comment créer un objet DataFrame à partir d'un dictionnaire de listes Python.
</p>
<p>
Dans cet exemple, nous utilisons le constructeur d'objet DataFrame pandas avec un dictionnaire de listes Python en tant qu'argument. La valeur de chaque entrée dans le dictionnaire est le nom de la colonne et les listes sont leurs valeurs.
</p>

In [3]:

data = {'Année': [2010, 2011, 2012, 2010, 2011, 2012, 2010, 2011, 2012],
        'Equipes': ['FCBarcelona', 'FCBarcelona', 'FCBarcelona', 'RMadrid', 'RMadrid', 'RMadrid', 'ValenciaCF',
                 'ValenciaCF', 'ValenciaCF'],
        'matches gagnés':   [30, 28, 32, 29, 32, 26, 21, 17, 19],
        'matches égalités':  [6, 7, 4, 5, 4, 7, 8, 10, 8],
        'matches perdus': [2, 3, 2, 4, 2, 5, 9, 11, 11]
        }
equipes = pd.DataFrame(data, columns = ['Année', 'Equipes', 'matches gagnés', 'matches égalités', 'matches perdus'])
equipes

Année      Equipes  matches gagnés  matches égalités  matches perdus
0   2010  FCBarcelona              30                 6               2
1   2011  FCBarcelona              28                 7               3
2   2012  FCBarcelona              32                 4               2
3   2010      RMadrid              29                 5               4
4   2011      RMadrid              32                 4               2
5   2012      RMadrid              26                 7               5
6   2010   ValenciaCF              21                 8               9
7   2011   ValenciaCF              17                10              11
8   2012   ValenciaCF              19                 8              11

<p>
Le résultat est une table où chaque entrée du dictionnaire est une colonne dans l'objet DataFrame. L'index de chaque ligne est créé automatiquement en prenant la position de ses éléments dans les listes d'entrée, en commençant par 0. Bien qu'il soit très facile de créer des DataFrames à partir de zéro (from scratch), la plupart du temps on est amené à importer des données dans une structure DataFrame.
</p>

## Utilisation de "données ouvertes (open data)" dans Pandas
<p>
Pour illustrer comment nous pouvons utiliser pandas dans un problème réel simple, nous allons prendre une base des données contenant des données publiques ouvertes. Les données ouvertes sont librement utilisées, réutilisées et distribuées par n'importe qui.
</p>
<p>
<a href="http://ec.europa.eu/eurostat/data/database">Eurostat</a> fournit plusieurs bases de données libres. Eurostat est une direction générale de la Commission européenne chargée de l'information statistique à l'échelle communautaire. Elle a pour rôle de produire les statistiques officielles de l'Union européenne, principalement en collectant, harmonisant et agrégeant les données publiées par les instituts nationaux de statistiques des pays membres de l'Union européenne, des pays candidats à l'adhésion et des pays de l'Association européenne de libre-échange.
</p>
<p>
Comme la quantité de données dans la base de données d'Eurostat est énorme, nous allons  nous concentrer seulement sur les données relatives aux indicateurs de financement de l'éducation par les États membres. Ainsi, la première chose à faire est de récupérer ces données auprès d'Eurostat. Étant donné que les données ouvertes doivent être fournies dans un format de texte brut, les formats CSV (ou tout autre format délimité par délimiteur) sont couramment utilisés pour stocker des données tabulaires. Dans un fichier de valeurs séparées par des délimiteurs, chaque ligne est un enregistrement de données et chaque enregistrement est constitué d'un ou plusieurs champs, séparés par le caractère délimiteur (généralement une virgule). 
</p>
<p>
Les données que nous allons utiliser dans la suite de ce chapitre peuvent être téléchargées et prétraitées dans un fichier CSV nommé <i>educ_figdp_1_Data.csv</i>, comme il peut être récupéré à partir du dossier file de ce repository.
</p>

## Lecture de données tabulaires dans pandas

<p>
Commençons par lire les données que nous avons téléchargées. La façon de lire les fichiers CSV dans Pandas consiste à utiliser la méthode <strong>read_csv</strong>. En plus du nom du fichier, nous ajoutons l'argument clé <strong>na_values</strong> à cette méthode avec le caractère représentant les "données non disponibles" dans le fichier. Normalement, les fichiers CSV ont un en-tête avec les noms des colonnes. Si tel est le cas, nous pouvons utiliser le paramètre <strong>usecols</strong> pour sélectionner les colonnes du fichier qui seront utilisées.
</p>

In [4]:
# remarquez que le fichier se trouve dans le dossier 'files' du repository
education = pd.read_csv('files/educ_figdp_1_Data.csv',
                  na_values=':', usecols=['TIME', 'GEO', 'Value'])
education

TIME                            GEO  Value
0    2000  European Union (28 countries)    NaN
1    2001  European Union (28 countries)    NaN
2    2002  European Union (28 countries)   5.00
3    2003  European Union (28 countries)   5.03
4    2004  European Union (28 countries)   4.95
5    2005  European Union (28 countries)   4.92
6    2006  European Union (28 countries)   4.91
7    2007  European Union (28 countries)   4.92
8    2008  European Union (28 countries)   5.04
9    2009  European Union (28 countries)   5.38
10   2010  European Union (28 countries)   5.41
11   2011  European Union (28 countries)   5.25
12   2000  European Union (27 countries)   4.91
13   2001  European Union (27 countries)   4.99
14   2002  European Union (27 countries)   5.00
15   2003  European Union (27 countries)   5.04
16   2004  European Union (27 countries)   4.95
17   2005  European Union (27 countries)   4.92
18   2006  European Union (27 countries)   4.91
19   2007  European Union (27 countries)   4.93
20   2008  European Union (27 countries)   5.04
21   2009  European Union (27 countries)   5.38
22   2010  European Union (27 countries)   5.41
23   2011  European Union (27 countries)   5.25
24   2000  European Union (25 countries)   4.94
25   2001  European Union (25 countries)   5.02
26   2002  European Union (25 countries)   5.03
27   2003  European Union (25 countries)   5.06
28   2004  European Union (25 countries)   4.98
29   2005  European Union (25 countries)   4.95
..    ...                            ...    ...
354  2006                       Slovenia   5.72
355  2007                       Slovenia   5.15
356  2008                       Slovenia   5.20
357  2009                       Slovenia   5.69
358  2010                       Slovenia   5.68
359  2011                       Slovenia   5.68
360  2000                       Slovakia   3.92
361  2001                       Slovakia   3.99
362  2002                       Slovakia   4.31
363  2003                       Slovakia   4.30
364  2004                       Slovakia   4.19
365  2005                       Slovakia   3.85
366  2006                       Slovakia   3.80
367  2007                       Slovakia   3.62
368  2008                       Slovakia   3.61
369  2009                       Slovakia   4.09
370  2010                       Slovakia   4.22
371  2011                       Slovakia   4.06
372  2000                        Finland   5.89
373  2001                        Finland   6.06
374  2002                        Finland   6.22
375  2003                        Finland   6.43
376  2004                        Finland   6.42
377  2005                        Finland   6.30
378  2006                        Finland   6.18
379  2007                        Finland   5.90
380  2008                        Finland   6.10
381  2009                        Finland   6.81
382  2010                        Finland   6.85
383  2011                        Finland   6.76

[384 rows x 3 columns]

<p>
Dans ce cas, le DataFrame résultant de la lecture de nos données est stocké dans l'objet <strong>education</strong>. La sortie de l'exécution montre que la taille du DataFrame est de 384 lignes X times 3 colonnes. Puisque le DataFrame est trop grand pour être entièrement affiché, trois points de suspensions apparaissent au milieu de chaque ligne.
</p>
<p>
En plus de la méthode <strong>read_csv</strong>, Pandas fournit aussi des fonctions pour lire des fichiers avec des formats comme Excel, HDF5, des fichiers tabulaires ou même le contenu du presse-papiers (<strong>read_excel ()</strong>, <strong>read_hdf ()</strong>, <strong>read_table ()</strong>, <strong>read_clipboard ()</strong>) . Quelle que soit la fonction utilisée, le résultat de la lecture d'un fichier est stocké en tant qu'objet DataFrame.
</p>

## Affichage des données
<p>
Pour voir à quoi ressemblent les données, nous pouvons utiliser la méthode <strong>head ()</strong>, qui montre seulement les cinq premières lignes. Si nous passons un nombre en argument de cette méthode, ce sera le nombre des premières lignes affichées.
</p>

In [5]:
# Affichage des 5 premières lignes
education.head() #Equivalent à education.head(5)

TIME                            GEO  Value
0  2000  European Union (28 countries)    NaN
1  2001  European Union (28 countries)    NaN
2  2002  European Union (28 countries)   5.00
3  2003  European Union (28 countries)   5.03
4  2004  European Union (28 countries)   4.95

In [6]:
# Affichage des 10 premières lignes
education.head(10)

TIME                            GEO  Value
0  2000  European Union (28 countries)    NaN
1  2001  European Union (28 countries)    NaN
2  2002  European Union (28 countries)   5.00
3  2003  European Union (28 countries)   5.03
4  2004  European Union (28 countries)   4.95
5  2005  European Union (28 countries)   4.92
6  2006  European Union (28 countries)   4.91
7  2007  European Union (28 countries)   4.92
8  2008  European Union (28 countries)   5.04
9  2009  European Union (28 countries)   5.38

De même, il existe la méthode <strong>tail ()</strong>, qui retourne les cinq dernières lignes par défaut.

In [7]:
education.tail()

TIME      GEO  Value
379  2007  Finland   5.90
380  2008  Finland   6.10
381  2009  Finland   6.81
382  2010  Finland   6.85
383  2011  Finland   6.76

In [8]:
education.tail(3)

TIME      GEO  Value
381  2009  Finland   6.81
382  2010  Finland   6.85
383  2011  Finland   6.76

<p>
Si nous voulons connaître les noms des colonnes ou les noms des index, nous pouvons utiliser les attributs <strong>columns</strong> et <strong>index</strong> de DataFrame. Les noms des colonnes ou des index peuvent être modifiés en affectant une nouvelle liste de la même longueur à ces attributs.
</p>

In [9]:
education.columns

Index(['TIME', 'GEO', 'Value'], dtype='object')

In [10]:
education.index

RangeIndex(start=0, stop=384, step=1)

<p>
Les valeurs d'un DataFrame peuvent être récupérées en tant que tableau Python en appelant son attribut values.
</p>

In [11]:
education.values

array([[2000, 'European Union (28 countries)', nan],
       [2001, 'European Union (28 countries)', nan],
       [2002, 'European Union (28 countries)', 5.0],
       ..., 
       [2009, 'Finland', 6.81],
       [2010, 'Finland', 6.85],
       [2011, 'Finland', 6.76]], dtype=object)

<p>
Si nous voulons simplement des informations statistiques rapides sur toutes les colonnes numériques d'un DataFrame, nous pouvons utiliser la fonction <strong>describe ()</strong>. Le résultat de la fonction montre le nombre, la moyenne, l'écart type, le minimum et le maximum, et les centiles, par défaut, les 25e, 50e et 75e, pour toutes les valeurs de chaque colonne (chaque série).
</p>

In [12]:
education.describe()

TIME       Value
count   384.000000  361.000000
mean   2005.500000    5.203989
std       3.456556    1.021694
min    2000.000000    2.880000
25%    2002.750000    4.620000
50%    2005.500000    5.060000
75%    2008.250000    5.660000
max    2011.000000    8.810000

## La sélection

<p>
Si nous voulons sélectionner un sous-ensemble de données à partir d'un DataFrame, il est nécessaire d'indiquer ce sous-ensemble à l'aide de crochets [] après le DataFrame. Le sous-ensemble peut être spécifié de plusieurs façons. Si nous voulons sélectionner seulement une colonne d'un DataFrame, nous avons  besoin seulement de mettre son nom entre les crochets. Le résultat sera une série, et non pas un DataFrame, car une seule colonne est récupérée.
</p>

In [13]:
# Sélectionner la colonne 'Value' du DataFrame
education['Value']

0       NaN
1       NaN
2      5.00
3      5.03
4      4.95
5      4.92
6      4.91
7      4.92
8      5.04
9      5.38
10     5.41
11     5.25
12     4.91
13     4.99
14     5.00
15     5.04
16     4.95
17     4.92
18     4.91
19     4.93
20     5.04
21     5.38
22     5.41
23     5.25
24     4.94
25     5.02
26     5.03
27     5.06
28     4.98
29     4.95
       ... 
354    5.72
355    5.15
356    5.20
357    5.69
358    5.68
359    5.68
360    3.92
361    3.99
362    4.31
363    4.30
364    4.19
365    3.85
366    3.80
367    3.62
368    3.61
369    4.09
370    4.22
371    4.06
372    5.89
373    6.06
374    6.22
375    6.43
376    6.42
377    6.30
378    6.18
379    5.90
380    6.10
381    6.81
382    6.85
383    6.76
Name: Value, Length: 384, dtype: float64

<p>
Si nous voulons sélectionner un sous-ensemble de lignes d'un DataFrame, nous pouvons le faire en indiquant une plage de lignes entre crochets et séparées par <strong>":"</strong>.
</p>
<p>
L'instruction suivante renvoie la tranche de lignes entre la 10ème à la 13ème position. Notez que la tranche n'utilise pas les étiquettes d'index comme références, mais la position. Dans ce cas, les étiquettes des lignes coïncident simplement avec la position des lignes.
</p>

In [14]:
education[10:14]

TIME                            GEO  Value
10  2010  European Union (28 countries)   5.41
11  2011  European Union (28 countries)   5.25
12  2000  European Union (27 countries)   4.91
13  2001  European Union (27 countries)   4.99

<p>
Si nous voulons sélectionner un sous-ensemble de colonnes et de lignes en utilisant les étiquettes comme références au lieu des positions, nous pouvons utiliser l'indexation <strong>loc</strong> :
</p>
<p>
L'instruction suivante renvoie toutes les lignes ayant :
<ul>
<li>les index spécifiés avant la virgule
<li>les colonnes spécifiées en tant que liste après la virgule
</ul>
<p>
Dans ce cas, <strong>loc</strong> ne renvoie pas les 90<sup>e</sup> à 94<sup>e</sup> lignes, mais renvoie toutes les lignes entre la ligne étiquetée 90 et la ligne étiquetée 94.
</p>

In [15]:
education.loc[90:94, ['TIME', 'GEO']]

TIME      GEO
90  2006  Belgium
91  2007  Belgium
92  2008  Belgium
93  2009  Belgium
94  2010  Belgium

## Filtrer les données
<p>
Une autre façon de sélectionner un sous-ensemble de données consiste à appliquer une condition booléenne. Cette condition est communément appelée un filtre. Par exemple, si nous voulons sélectionner seulement les 5 premières lignes ayant la colonne 'Value' inférieures ou égales à 6.5, nous pouvons le faire comme ceci :
</p>

In [16]:
#education[education['Value'] > 6.5] affiche toutes les lignes ayant 'Value' <= 6.5
education[education['Value'] <= 6.5].head()

TIME                            GEO  Value
2  2002  European Union (28 countries)   5.00
3  2003  European Union (28 countries)   5.03
4  2004  European Union (28 countries)   4.95
5  2005  European Union (28 countries)   4.92
6  2006  European Union (28 countries)   4.91

<p>
La condition booléenne renvoie <strong>True</strong> ou <strong>False</strong> pour chaque ligne. Les lignes qui correspondent à la valeur True seront sélectionnées.
</p>
<p>
Dans l'exemple précédent, l'opération booléenne <strong>education ['Value'] &lt;= 6.5</strong> produit un masque booléen. Lorsqu'un élément de la colonne 'Value' est inférieur ou égal à 6.5, la valeur correspondante dans le masque est définie par True, sinon elle est définie par False.
</p>
<p>
Bien sûr, les opérateurs booléens habituels peuvent être utilisés pour le filtrage :
</p>
<table border="1">
<tr>
<td>&lt;</td><td>inférieur à</td>
</tr>
<tr>
<td> &lt;=</td><td>inférieur ou égal à</td>
</tr>
<tr>
<td>&gt;</td><td>supérieur à</td>
</tr>
<tr>
<td>&gt;=</td><td>supérieur ou égal à</td>
</tr>
<tr>
<td>=</td><td>égal à</td>
</tr>
<tr>
<td>!=</td><td>différent de</td>
</tr>
</table>


## Filtrage des valeurs manquantes
<p>
Pandas utilise la valeur spéciale <strong>NaN</strong> (Not a Number, pas un nombre) pour représenter les valeurs manquantes. Une caractéristique subtile des valeurs NaN est que deux NaN ne sont jamais égaux. Pour cette raison, le seul moyen sûr de savoir si une valeur est manquante dans un DataFrame est d'utiliser la fonction <strong>isnull()</strong> :
</p>

In [17]:
education[education['Value'].isnull()].head()

TIME                            GEO  Value
0   2000  European Union (28 countries)    NaN
1   2001  European Union (28 countries)    NaN
36  2000       Euro area (18 countries)    NaN
37  2001       Euro area (18 countries)    NaN
48  2000       Euro area (17 countries)    NaN

## Manipulation des données

<p>
Une fois que nous savons comment sélectionner les données souhaitées, la prochaine étape est de savoir comment manipuler les données. L'une des choses les plus simples que nous puissions faire est de manipuler des colonnes ou des lignes en utilisant des fonctions d'agrégation. La liste suivante montre les fonctions d'agrégation les plus courantes :
</p>

<table>
<tr>
<td>Fonction</td><td>Description</td>
</tr>
<tr>
<td>count()</td><td>Nombres d'observations non nulles</td>
</tr>
<tr>
<td>sum()</td><td>La somme des valeurs</td>
</tr>
<tr>
<td>mean()</td><td>La moyenne des valeurs</td>
</tr>
<tr>
<td>median()</td><td>La médiane arithmétique des valeurs</td>
</tr>
<tr>
<td>min()</td><td>Le minimum de plusieurs valeurs</td>
</tr>
<tr>
<td>max()</td><td>Le maximum de plusieurs valeurs</td>
</tr>
<tr>
<td>prod()</td><td>Le produits de plusieurs valeurs</td>
</tr>
<tr>
<td>std()</td><td>Ladéviation standard</td>
</tr>
<tr>
<td>var()</td><td>La variance</td>
</tr>
</table>

<p>
Le résultat de toutes ces fonctions appliquées à une ligne ou à une colonne est toujours un nombre. Si une fonction est appliquée à un DataFrame ou à une sélection de lignes et de colonnes, vous pouvez spécifier si la fonction doit être appliquée aux lignes de chaque colonne (en plaçant le mot-clé <strong>axis = 0</strong> dans l'appel de la fonction), ou aux colonnes de chaque ligne (en mettant le mot-clé <strong>axis = 1</strong> dans l'appel de la fonction).
</p>

In [18]:
education.max(axis=0)

TIME      2011
GEO      Spain
Value     8.81
dtype: object

In [19]:
education.sum(axis=1).head()

0    2000.00
1    2001.00
2    2007.00
3    2008.03
4    2008.95
dtype: float64

<p>
Notez que les fonctions citées ci-dessus sont spécifiques Pandas, et ne s'agissent pas de fonctions de Python. Il y a des différences dans leur mise en œuvre. En Python, les valeurs NaN sont prises en comptes par les opérations sans déclencher d'exception. En revanche, les opérations Pandas excluent les valeurs NaN représentant les données manquantes. Par exemple, la fonction pandas max() exclut les valeurs NaN, elles sont donc interprétées comme des valeurs manquantes, tandis que la fonction Python max() prendra NAN comme maximum :
</p>

In [20]:
print ('Fonction max de Pandas :', education['Value'].max())
print ('Fonction max Python  :', max(education['Value']))

Fonction max de Pandas : 8.81
Fonction max Python  : nan


<p>
En plus de ces fonctions d'agrégation, nous pouvons appliquer des opérations sur toutes les valeurs des lignes, des colonnes ou une sélection des deux. La règle générale est qu'une opération sur les colonnes signifie qu'elle est appliquée à chaque ligne de cette colonne et qu'une opération sur les lignes signifie qu'elle est appliquée à chaque colonne de cette ligne. Par exemple, nous pouvons appliquer n'importe quelle opération arithmétique binaire (+, -, *, /) à une ligne ou colonne entière :
</p> 

In [21]:
value = education['Value']
value.head()/10

0      NaN
1      NaN
2    0.500
3    0.503
4    0.495
Name: Value, dtype: float64

<p>
Nous pouvons appliquer, aussi, n'importe quelle fonction à un DataFrame ou à une série en plaçant simplement son nom comme argument de la méthode <strong>apply()</strong>. Par exemple, dans le code suivant, nous appliquons la fonction <strong>sqrt()</strong> de la bibliothèque <strong>numpy</strong> pour calculer la racine carrée de chaque valeur dans la colonne 'Value'.
</p>

In [22]:
s = education['Value'].apply(np.sqrt)
s.head()

0         NaN
1         NaN
2    2.236068
3    2.242766
4    2.224860
Name: Value, dtype: float64

<p>
Une autre manipulation de base consiste à définir de nouvelles valeurs dans notre DataFrame. Cela peut être fait directement en utilisant l'opérateur d'affectation <strong>=</strong> sur un DataFrame. Par exemple, pour ajouter une nouvelle colonne à un DataFrame, nous pouvons affecter une série à une colonne qui n'existe pas. Cela produira une nouvelle colonne dans le DataFrame après tous les autres. Vous devez être conscient que si une colonne avec le même nom existe déjà, les valeurs précédentes seront remplacées.
</p>
<p>
Dans l'exemple suivant, nous affectons la série qui résulte de la division de la colonne "Value" par la valeur maximale de la même colonne à une nouvelle colonne nommée "ValueNorm".
</p>

In [23]:
education['ValueNorm'] = education['Value'] / education['Value'].max()
education.tail()

TIME      GEO  Value  ValueNorm
379  2007  Finland   5.90   0.669694
380  2008  Finland   6.10   0.692395
381  2009  Finland   6.81   0.772985
382  2010  Finland   6.85   0.777526
383  2011  Finland   6.76   0.767310

<p>
Maintenant, si nous voulons supprimer une colonne du DataFrame, nous pouvons utiliser la fonction <strong>drop</strong>; cela supprime les lignes indiquées si <strong>axis = 0</strong>, ou les colonnes indiquées si <strong>axis = 1</strong>.
</p>
<p>
Dans Pandas, toutes les fonctions qui modifient le contenu d'un DataFrame, comme la fonction <strong>drop</strong>, renvoient normalement une copie des données modifiées, au lieu d'écraser le DataFrame. Par conséquent, le DataFrame d'origine est conservé. Si vous ne souhaitez pas conserver les anciennes valeurs, vous pouvez définir le mot clé <strong>inplace</strong> à True. Par défaut, ce mot clé est défini à False, ce qui signifie qu'une copie des données est renvoyée.
</p>

In [24]:
education.drop('ValueNorm', axis=1)
education.head()

TIME                            GEO  Value  ValueNorm
0  2000  European Union (28 countries)    NaN        NaN
1  2001  European Union (28 countries)    NaN        NaN
2  2002  European Union (28 countries)   5.00   0.567537
3  2003  European Union (28 countries)   5.03   0.570942
4  2004  European Union (28 countries)   4.95   0.561862

In [25]:
education.drop('ValueNorm', axis=1, inplace=True)
education.head()

TIME                            GEO  Value
0  2000  European Union (28 countries)    NaN
1  2001  European Union (28 countries)    NaN
2  2002  European Union (28 countries)   5.00
3  2003  European Union (28 countries)   5.03
4  2004  European Union (28 countries)   4.95

<p>
Pour insérer une nouvelle ligne en bas du DataFrame, nous pouvons utiliser la fonction Pandas <strong>append()</strong>. Cette fonction reçoit comme argument la nouvelle ligne, qui est représentée comme un dictionnaire où les clés sont le nom des colonnes et les valeurs la valeur associée. Vous devez mettre l'indicateur <strong>ignore_index</strong> dans la méthode append à True, sinon l'index 0 est donné à cette nouvelle ligne, ce qui produira une erreur si cet index existe déjà :
</p>

In [26]:
education = education.append({'TIME': 2000, 'Value': 5.00, 'GEO': 'a'}, ignore_index=True)
education.tail()

TIME      GEO  Value
380  2008  Finland   6.10
381  2009  Finland   6.81
382  2010  Finland   6.85
383  2011  Finland   6.76
384  2000        a   5.00

<p>
Si nous voulons supprimer cette ligne, nous devons à nouveau utiliser la fonction <strong>drop</strong>. Cette fois-ci, nous devons initier l'argument <strong>axis</strong> avec 0 ( <strong>axis = 0</strong>), et spécifier l'index de la ligne que nous voulons supprimer. Puisque nous voulons supprimer la dernière ligne, nous pouvons utiliser la fonction <strong>max()</strong> sur les index pour déterminer de quelle ligne s'agit-il.
</p>

In [27]:
education.drop(max(education.index), axis=0, inplace=True)
education.tail()

TIME      GEO  Value
379  2007  Finland   5.90
380  2008  Finland   6.10
381  2009  Finland   6.81
382  2010  Finland   6.85
383  2011  Finland   6.76

<p>
La fonction <strong>drop ()</strong> est également utilisée pour supprimer les valeurs manquantes en l'appliquant sur le résultat de la fonction <strong>isnull ()</strong>. Cela a un effet similaire à filtrer les valeurs <trong>NaN</strong>, comme nous l'avons expliqué ci-dessus, mais ici la différence est qu'une copie du DataFrame sans les valeurs <strong>NaN</strong> est retournée, au lieu d'une vue.
</p>

In [28]:
eduNaN = education.drop(education['Value'].isnull(), axis=0)
eduNaN.head()

TIME                            GEO  Value
2  2002  European Union (28 countries)   5.00
3  2003  European Union (28 countries)   5.03
4  2004  European Union (28 countries)   4.95
5  2005  European Union (28 countries)   4.92
6  2006  European Union (28 countries)   4.91

<p>
Pour supprimer les valeurs <strong>NaN</strong>, au lieu d'utiliser la fonction de suppression générique (<strong>drop()</strong>), nous pouvons utiliser la fonction spécifique <strong>dropna()</strong>. Si nous voulons effacer une ligne contenant une valeur NaN, nous devons affecter au mot-clé <strong>how</strong> la valeur <strong>any</strong>.
Pour le restreindre à un sous-ensemble de colonnes, nous pouvons le spécifier en utilisant le mot clé <strong>subset</strong>. Comme nous pouvons le voir ci-dessous, le résultat sera identique à l'utilisation de la fonction <strong>drop</strong> :
</p>

In [29]:
eduNaN = education.dropna(how='any', subset=['Value'], axis=0)
eduNaN.head()

TIME                            GEO  Value
2  2002  European Union (28 countries)   5.00
3  2003  European Union (28 countries)   5.03
4  2004  European Union (28 countries)   4.95
5  2005  European Union (28 countries)   4.92
6  2006  European Union (28 countries)   4.91

<p>
Si, au lieu de supprimer les lignes contenant NaN, nous voulons les remplir avec une autre valeur, alors nous pouvons utiliser la méthode <strong>fillna()</strong>, en spécifiant quelle valeur doit être utilisée. Si nous voulons remplir seulement certaines colonnes spécifiques, nous devons mettre comme argument de la fonction <strong>fillna()</strong> un dictionnaire avec le nom des colonnes comme clé et le caractère à utiliser pour le remplissage comme valeur.
</p>

In [31]:
eduFilled = education.fillna(value={'Value': 0})
eduFilled.head()

TIME                            GEO  Value
0  2000  European Union (28 countries)   0.00
1  2001  European Union (28 countries)   0.00
2  2002  European Union (28 countries)   5.00
3  2003  European Union (28 countries)   5.03
4  2004  European Union (28 countries)   4.95

## Tri de données

<p>
Une autre fonctionnalité importante dont nous aurons besoin lors de l'inspection de nos données est de trier les données par colonnes. Nous pouvons trier un DataFrame en utilisant n'importe quelle colonne, en utilisant la fonction de tri (<strong>sort_xxxx()</strong>). Si nous voulons voir les cinq premières lignes des données triées par ordre décroissant en utilisant la colonne <strong>'Value'</strong>, alors nous pouvons écrire :
</p>

In [32]:
education.sort_values(by='Value', ascending=False, inplace=True)
education.head()

TIME      GEO  Value
130  2010  Denmark   8.81
131  2011  Denmark   8.75
129  2009  Denmark   8.74
121  2001  Denmark   8.44
122  2002  Denmark   8.44

<p>
Notez que le mot-clé <strong>inplace</strong> signifie que le DataFrame sera écrasé, et donc aucun nouveau DataFrame n'est retourné. Si au lieu de <strong>ascending = False</strong>, nous utilisons <strong>ascending = True</strong>, les valeurs seront triées dans l'ordre croissant.
</p>
<p>
Si nous voulons revenir à l'ordre d'origine, nous pouvons trier par un index en utilisant la fonction <strong>sort_index</strong> et en spécifiant <strong>axis = 0</strong> :
</p>

In [33]:
education.sort_index(axis=0, ascending=True, inplace=True)
education.head()

TIME                            GEO  Value
0  2000  European Union (28 countries)    NaN
1  2001  European Union (28 countries)    NaN
2  2002  European Union (28 countries)   5.00
3  2003  European Union (28 countries)   5.03
4  2004  European Union (28 countries)   4.95

## Regroupement des données

<p>
Un autre moyen très utile pour inspecter les données consiste à les regrouper selon certains critères. Par exemple, dans notre cas, il serait bien judicieux de regrouper toutes les données par pays, indépendamment de l'année. Pandas dispose de la fonction <strong>groupby</strong> qui nous permet de faire exactement cela. La valeur renvoyée par cette fonction est un DataFrame dans lequel les valeurs sont regroupés. Pour avoir un DataFrame correct, il est nécessaire d'appliquer une fonction d'agrégation (par exemple, sum(), mean, ...). Ainsi, cette fonction sera appliquée à toutes les valeurs du même groupe.
</p>
<p>
Par exemple, dans notre cas, si nous voulons un DataFrame affichant la moyenne des valeurs pour chaque pays pour toutes les années, nous pouvons l'obtenir en regroupant par pays et en utilisant la fonction <strong>mean</strong> comme méthode d'agrégation pour chaque groupe. Le résultat serait un DataFrame avec des pays comme index et les valeurs des moyennes comme colonnes :
</p>

In [36]:
group = education[['GEO', 'Value']].groupby('GEO').mean()
group.head()

Value
GEO                     
Austria         5.618333
Belgium         6.189091
Bulgaria        4.093333
Cyprus          7.023333
Czech Republic  4.168333